Drop CI support for Ruby < 2.3

pirj opened the pull request:
Extracted from:
 - https://github.com/rspec/rspec-core/pull/2787
 - https://github.com/rspec/rspec-support/pull/436
 - https://github.com/rspec/rspec-mocks/pull/1349
 - https://github.com/rspec/rspec-expectations/pull/1231

Repo PRs:
 - https://github.com/rspec/rspec-mocks/pull/1358
 - https://github.com/rspec/rspec-core/pull/2790
 - https://github.com/rspec/rspec-support/pull/444
 - https://github.com/rspec/rspec-expectations/pull/1234

**NOT DONE** (mostly JRuby-related, saved for a subsequent iteration):
 -    Remove `continue-on-error: ${{ matrix.ruby == 'jruby-9.2.13.0' || endsWith(matrix.ruby, 'head') }}`
 -    add jruby-9.1, 9.2 in `ci.yml` (bumps 9.1.7.0 to 9.1.17.0)
 -    remove jruby from travis
 -    always use bundler 2         `bundler: ${{ (matrix.ruby == 'jruby-9.1.17.0' && 1) || 2 }}`
 -    `# For some reason JRuby doesn't like our improved bundler setup`
 -    `yes | gem update --system --no-document` - JRuby fails to install docs sometimes
 -    `yes | gem install bundler --no-document`
 -    turn on for all rubies? `function documentation_enforced {`/`if is_mri; then`

pirj commented:
@JonRowe @benoittgt Please take a look.
I've also removed `appveyor.yml` along with all its traces in pull requests.

pirj commented:
No idea why Travis `DIFF_LCS` builds don't pick up the right branch of `rspec-rails` to clone or `RSPEC_CI=true` trick doesn't work:
```
Bundler could not find compatible versions for gem "rspec-core":

  In Gemfile:
    rspec-core
    rspec-rails was resolved to 4.1.0.pre, which depends on
      rspec-core (= 3.11.0.pre)

Could not find gem 'rspec-core (= 3.11.0.pre)', which is required by gem 'rspec-rails', in any of the relevant sources:
  source at `/home/travis/build/rspec/rspec-core`
```

Don't think it makes much sense to fix it here.
I'd rather remove the remaining of Travis builds, hope you guys had a chance at moving `DIFF_LCS` and Rubinius builds over to GHA.

pirj commented on ci/script/clone_all_rspec_repos:
```diff
@@ -9,11 +9,8 @@ if is_mri; then
-  if rspec_support_compatible; then
-    clone_repo "rspec-support"
-  fi
+  clone_repo "rspec-rails" "4-0-dev"
```
Maybe we should introduce a `maintenance-rspec-rails-branch` file?

Still, that `4-0-dev` branch has nothing to do with `rspec-rails` 4.0, but rather it's there to adjust `rspec-rails` dependencies so they point back to the right version of the current repository where it's tested in a sub-build.

JonRowe replied:
The `rspec-rails` branch you're looking for is `main` for `4-0-dev`, if you'd like to build out the file based approach and use it for both branches I'd be happy with that.

pirj replied:
`main` doesn't work, because it [has `main`](https://github.com/rspec/rspec-rails/blob/9217638aff96df56753eeafe60e16bc4c35b8566/maintenance-branch#L1) in it, this means that other `rspec-*` projects will be used for `rspec-rails` sub build. `4-0-dev` [points to](https://github.com/rspec/rspec-rails/blob/69e23e20166f704a52e30d9b1a3a33d608def353/maintenance-branch#L1) `4-0-dev` branch, and `rspec-*` will be used from it.

I think we can live with temporarily having a `4-0-dev` branch in `rspec-rails` and reverting this line here when we do a release and `4-0-dev` is merged to `main`.

JonRowe replied:
> main doesn't work, because it has main in it, this means that other rspec-* projects will be used for rspec-rails sub build. 4-0-dev points to 4-0-dev branch, and rspec-* will be used from it.

No each project has `4-0-dev` in and points to `rspec-rails` main, thats what this function does, it overrides the `4-0-dev` in the local file.

pirj replied:
Oh, I see! It doesn't really matter what is in `maintenance-branch` of the cloned `rspec-rails`, it's unused in sub-builds.
We check [if there are necessary repos in sibling directories](
https://github.com/rspec/rspec-rails/blob/9217638aff96df56753eeafe60e16bc4c35b8566/Gemfile-rspec-dependencies#L4) and use them when bundling.
This slipped my mind completely.

There was a strange build failure that `bundler` cannot find `rspec-rails`'s `4.0.0.pre` `rspec-support` dependency in any of the sources. Might have been temporary.

I intend to:
 - delete `4-0-dev` `rspec-rails` branch
 - change to `clone_repo "rspec-rails" "main"` here and in generated PRs

Thanks for the heads-up 👍

pirj replied:
Found it. The failure was because `RSPEC_CI: true` was not set.

JonRowe commented on ci/.travis.yml:
```diff
@@ -12,21 +12,14 @@ before_install:
   include:
-    - rvm: jruby-1.7
-      env: JRUBY_OPTS='--dev --1.8'
     - rvm: 2.7.1
```
You can remove the diff lcs version builds too, and set the minimum diff lcs version.

pirj replied:
Should I use `">= 1.4.3", "< 2.0"`?
I'll make those changes separately.

JonRowe replied:
`>= 1.4.4`

JonRowe replied:
Can you just add it as a commit to the prs.

pirj replied:
Sure, but which one? 😄 I have those four generated from this PR, also those ones from which I've extracted these changes.

I plan to send a few more series:
 - update rake to 13, and cucumber to 2.4 (broke my teeth trying to run 1.3 locally) <--- **will add diff-lcs update here**
 - rethink `define_optimized_require_for_rspec`
 - fix kwargs delegation
 - update JRuby to 9.1.17.0/ add JRuby 9.2
 - get rid of most of the `<<-`/`chomp`/`unindent`/`undent`/`gsub` in favour of `<<~`

Do you have some suggestions for this plan?

JonRowe replied:
> update rake to 13, and cucumber to 2.4 (broke my teeth trying to run 1.3 locally) <--- will add diff-lcs update here

Please just add diff-lcs to these PRs, any gem which mentions it in the gemspec just set the minimum version and remove the diff lcs builds.

> *rethink `define_optimized_require_for_rspec`

I want to keep this, it is a lot of churn just to remove a generated method.

> fix kwargs delegation

We may find this is impossible due to the hash behaviour, I suspect we will need to keep ruby 2 keywords for quite some time.

> get rid of most of the <<-/chomp/unindent/undent/gsub in favour of <<~

That can wait until later prs, its subjective please don't combine it into your mega PRs they are already too hard to review.

> update JRuby to 9.1.17.0/ add JRuby 9.2

Do this seperatly as I'm also working on this for main

pirj replied:
> I plan to send a few more series

I'd like to emphasize and elaborate on the above statement ☝️ to remove all the ambiguity.
I plan to send series (dev+support+core+mocks+expectations) of PRs for each of those changes separately.
After this current one is merged.
All against `4-0-dev`.

>> define_optimized_require_for_rspec

> I want to keep this

👍 

> Please just add diff-lcs to these PRs, any gem which mentions it in the gemspec just set the minimum version and remove the diff lcs builds.

Will do in a follow-up 👍 

>> fix kwargs delegation

> this is impossible due to the hash behaviour, I suspect we will need to keep ruby 2 keywords for quite some time.

That's is the plan, to use `ruby2_keywords`, there's no other way around. I intend to use [this comment](https://github.com/rspec/rspec-expectations/issues/1221#issuecomment-716133588) as the guideline.

JonRowe commented on ci/.travis.yml:
```diff
@@ -12,21 +12,14 @@ before_install:
-  - 2.0.0
-  - ree
   - rbx-3
   - jruby-9.1.7.0 # pin JRuby to this until travis/rvm can install later versions
```
You can remove RBX-3

JonRowe commented on ci/script/functions.sh:
```diff
@@ -9,14 +9,20 @@ SPECS_HAVE_RUN_FILE=specs.out
 MAINTENANCE_BRANCH=`cat maintenance-branch`
 
 # Don't allow rubygems to pollute what's loaded. Also, things boot faster
```
Can we revert all these changes for now, it makes diffing harder and its easy to clean up post release.

pirj replied:
With an exception of `BRANCH_TO_CLONE` trick, we can. But we equally can just set `s.required_ruby_version = '>= 2.3.0'` and clean everything else after the release 😆 

Jokes aside, `rspec-mock` has undertaken [the smallest change, _just_  +360 −880](https://github.com/rspec/rspec-mocks/pull/1349) of all repos. Not much compared to [+45 -145 induced by this update](https://github.com/rspec/rspec-mocks/pull/1355).

All of the changes to `ci/script/functions.sh` and `ci/script/predicate_functions.sh` consist of removals and simple changes a-la `is_mri_2plus` -> `is_mri`. That can be skimmed through really quickly during the review (of the total `4-0-dev` -> `main` merge I suppose).

Except for the [`BRANCH_TO_CLONE` trick](https://github.com/rspec/rspec-core/commit/6c657734123171aab1811c7926c563bbd27b8013#diff-f686af268076eb6078295d47e8d97ab3df9486f0f22b900f4c045b5b9cc04fe7R22) that is in `rspec-*` repos' `4-0-dev`, but not on `main`, and not here, in `rspec-dev` anywhere. I've added it to this PR to avoid it being overwritten with `rake ci:update_files` next time we update `4-0-dev`.

JonRowe commented on ci/script/predicate_functions.sh:
```diff
@@ -1,67 +1,21 @@
 function is_mri {
-  if ruby -e "exit(!defined?(RUBY_ENGINE) || RUBY_ENGINE == 'ruby')"; then
```
Can we revert all these changes for now, it makes diffing harder and its easy to clean up post release.

JonRowe commented on ci/script/update_rubygems_and_install_bundler:
```diff
@@ -2,11 +2,5 @@
-  gem install bundler -v '1.17.3'
-fi
+yes | gem update --system
+yes | gem install bundler
```
Can we revert all these changes for now, it makes diffing harder and its easy to clean up post release.

JonRowe reviewed (changes requested):
Hi @pirj good work here, just some changes to drop more builds and a request to revert some changes for easy of diffing until after release

pirj commented:
I would leave this PR hanging until we release 4.0.
However, in those four generated PRs I've also cleaned up Appveyor, and I don't feel too amused when thinking that I'll have to re-do this.
It would be wrong to merge those four PRs without merging this one that generated them.
Also, this work has been done. I understand that it's hard work to review the changes, sometimes even more time-consuming than "delete function" hasty removals.
There are also four other pull requests from which I've extracted these changes, waiting. And I'll have to rearrange the numerous commits there to revert the changes to those to sh files. It's not like I wouldn't have to rearrange commits there otherwise, but it feels like some additional work to do.

To sum it up, it feels like a lot of extra work upfront, and also re-doing of the same removals I've done here at a later point in time. I don't feel too enthusiastic about doing that that compared to the overhead of reviewing this.

pirj commented on ci/.github/workflows/ci.yml:
```diff
@@ -8,6 +8,8 @@ on:
     branches:
       - '*'
+env:
+  RSPEC_CI: true
```
Forces `rspec-rails` to depend on `4.0.0.pre`. See https://github.com/rspec/rspec-rails/blob/934276b0e12b21ad44bac54494c42f5a93e08a33/rspec-rails.gemspec#L48

pirj commented on Rakefile:
```diff
@@ -383,7 +383,8 @@ namespace :ci do
   task :update_files do
-    update_ci_files_in_repos
+    opts = { except: %w[ rspec-rails ] }
+    update_ci_files_in_repos(opts)
```
Do not send CI file updates to `rspec-rails`.

pirj commented on ci/script/clone_all_rspec_repos:
```diff
@@ -9,11 +9,8 @@ if is_mri; then
   clone_repo "rspec-mocks"
-  clone_repo "rspec-rails"
-
-  if rspec_support_compatible; then
```
All builds are rspec-support compatible.

pirj commented on ci/script/clone_all_rspec_repos:
```diff
@@ -9,11 +9,8 @@ if is_mri; then
-  if rspec_support_compatible; then
-    clone_repo "rspec-support"
-  fi
+  clone_repo "rspec-rails" "main"
```
Not all builds (Ruby 2.3, Ruby 2.4) builds are `rspec-rails` compatible.
It's an unnecessary step to check it out for those builds, but I'll leave this improvement for later.

pirj commented on ci/script/functions.sh:
```diff
@@ -9,14 +9,20 @@ SPECS_HAVE_RUN_FILE=specs.out
   if [ ! -d $1 ]; then # don't clone if the dir is already there
-    travis_retry eval "git clone https://github.com/rspec/$1 --depth 1 --branch $MAINTENANCE_BRANCH"
+    if [ -z "$2" ]; then
+      BRANCH_TO_CLONE="$MAINTENANCE_BRANCH"
```
Because `rspec-rails` is alienated from other repos, there's no `4-0-dev` branch there, and we clone `main`.

pirj reviewed (commented):
Some notes to the future self.

JonRowe commented on ci/.travis.yml:
```diff
@@ -12,29 +12,14 @@ before_install:
-  - jruby-1.7
 env:
   - JRUBY_OPTS='--dev'
+  - RSPEC_CI='true'
```
You don't need this, its only used by Github to recognise windows CI.

pirj replied:
As far as I remember, `rspec-rails` sub-build breaks with https://github.com/rspec/rspec-dev/pull/259#discussion_r534432772

I'd rather keep this line since probably we'll get rid of this file altogether soon anyway.

JonRowe commented on ci/.travis.yml:
```diff
@@ -12,29 +12,14 @@ before_install:
 bundler_args: "--binstubs --standalone --without documentation --path ../bundle"
```
To be honest you can delete this file, we won't be using it after the release of RSpec 4, and Jruby is being worked on elsewhere.

pirj replied:
I could, but https://github.com/rspec/rspec-dev/blob/d27313db806a4a033208e4ac6f637f9009417bfb/ci/script/functions.sh#L82

```
bundle_install_flags=`cat .travis.yml | grep bundler_args | tr -d '"' | grep -o " .*"`
```

I tried to keep this PR focused to its title, "Drop CI support for Ruby < 2.3", not to upset code reviewers 😄 

Ok to adjust `ci/script/functions.sh` and inline those bundler flags? They don't differ across repos `"--binstubs --standalone --without documentation --path ../bundle"`.

I'd rather keep `.travis.yml` for now if you don't mind. Because some specs pass on all MRIs, but fail on JRuby on CI. And if my memory doesn't let me down, I couldn't reproduce them locally.
Again, `.travis.yml` will be removed soon anyway?

pirj commented:
I'll provide a short summary of what we've been discussing above in the comments.

 - [ ] follow-up: update rake to 13, cucumber to 2.4, will add diff-lcs to `>= 1.4.4`
 - [-] ~rethink define_optimized_require_for_rspec~
 - [ ] follow-up: fix kwargs delegation (with `ruby2_keywords`)
 - [ ] follow-up, probably wait for `main` and merge it in 4-0-dev: update JRuby to 9.1.17.0/ add JRuby 9.2
 - [ ] drop `.travis.ym`: can't do, I need JRuby builds for now
 - [ ] revert most of the changes to `script/functions.sh`/`script/predicate_functions.sh`/`script/update_rubygems_and_install_bundler` - I'm personally not sold on the tradeoff of redoing vs code review overhead
 - [ ] remove `RSPEC_CI` - `rspec-rails` sub-builds on Travis will break (JRuby are the only ones left)
 - [ ] inline `bundle_install_flags` to remove dependency of `script/functions.sh` on the presence of `.travis.yml`
 - [ ] follow-up (the last, since least important): get rid of most of the <<-/chomp/unindent/undent/gsub in favour of <<~
 - [ ] follow-up (post-4.0?): clean up `travis` naming in build functions

@JonRowe Do you want me to squeeze in some of the above changes in **this** PR, or ok with addressing in follow-ups?
Is this PR (and related generated PRs) good to merge as is?

pirj commented:
Rebased 4-0-dev on `main`, and this PR onto 4-0-dev.
Incoming update from `main` made it possible to remove `.travis.yml` (no Travis builds left!).

I'll send generated pulls when https://github.com/rspec/rspec-core/pull/2785 is green - `rspec-core`'s 4-0-dev is not yet in line with other repos in terms of CI matrices.
